### PR TITLE
Add an assertion and fix a failure of the assertion. 

### PR DESCRIPTION
**Pull request by vext01**
See commit messages for more info.

Companion to https://github.com/ykjit/ykllvm/pull/29

### CODE REVIEW

**Comment by ltratt** [comment]
Should there be a test for this?

**Comment by vext01** [comment]
I should have mentioned. The new assertion on the yk side was failing with existing tests.

**Comment by ltratt** [comment]
Aha, OK, let's get that into a commit message during squashing. Otherwise OK to me.

**Comment by vext01** [comment]
Agreed. I'd also like @ptersilie to give this the once-over.

File: tests/langtest_c.rs [hunk 1]
```diff
@@ -119,6 +119,7 @@ fn mk_compiler(exe: &Path, src: &Path, opt: &str, extra_objs: &[PathBuf]) -> Com
         "-Wl,--mllvm=--disable-early-taildup", // Interferes with the BlockDisambiguate pass.
         "-Wl,--mllvm=--disable-tail-duplicate", // ^^^
         "-Wl,--mllvm=--yk-disable-tail-call-codegen", // Interferes with the JIT's inlining stack.
+        "-Wl,--mllvm=--yk-no-fallthrough",     // Distorts the block mapping.
```
Comment (ptersilie) [inline tests/langtest_c.rs:125]: This flag doesn't distort the block mapping? Not using it distorts the mapping. So it un-distorts the block mapping.

**Comment by ptersilie** [comment]
This looks like this would also be solved with the "make function calls terminate blocks" pass we did once but never merged. I wonder if it's worth revisiting that at some point.

Apart from a "the the" typo in the first commit, looks good.

**Comment by ltratt** [comment]
> This looks like this would also be solved with the "make function calls terminate blocks" pass

Is that a more general / better solution?

**Comment by ptersilie** [comment]
I'm not sure. That's something we'd have to work out. Might be worth discussing with @vext01 offline at some point.

**Comment by vext01** [comment]
> This looks like this would also be solved with the "make function calls terminate blocks" pass we did once but never merged.

Do you mean make the `CallInstr` a terminator, or split each block after a call with an unconditional branch?

The former is a large change for the LLVM language. The latter doesn't address the underlying issue: that we may or may not resume from a call in the same block or another block.

**Comment by ptersilie** [comment]
I meant the latter, but I think I see now that this doesn't solve the issue.

**Comment by vext01** [comment]
> I meant the latter, but I think I see now that this doesn't solve the issue.

The problem case is as follows:
```
bbN:
  call f()
  br bbM

bbM:
  ...
```

And we may or may not resume in `BBM` after the call to `f`.

The call in `BBN` is already the last non-terminating instruction, so presumably the pass you are proposing would not change the block at all.

**Comment by ptersilie** [comment]
Yeah, I can't argue with that.

**Comment by vext01** [comment]
OK to squash?

**Comment by ptersilie** [comment]
Please do.

**Comment by ptersilie** [comment]
Oh wait, there's still one comment.

File: tests/langtest_c.rs (hunk 1 above)
Comment (vext01) [reply tests/langtest_c.rs]: fixed in [d142b30](https://github.com/ykjit/yk/pull/533/commits/d142b30250c2aa94dc7b203adc200c20630a1557)

**Comment by ptersilie** [comment]
Please squash.

**Comment by vext01** [comment]
splat.

**Comment by vext01** [comment]
bors r-

My bad. I was supposed to update the commit message.

**Comment by vext01** [comment]
In the first commit, the paragraph starting "For now we first add an assertion..." was reworded as per @ltratt's request.

If it looks good, please re-run bors @ptersilie.

**Pull request merged**